### PR TITLE
Enforce abstract class rule

### PR DESCRIPTION
**Pull request by veblush**
Abstract class cannot be instantiated but there are some cases to do that because it hasn't been strictly enforced by a compiler. This is because `GRPC_ABSTRACT` has been used instead of `= 0`. As a result, some classes with pure virtual functions happen to be instantiated even though it's actually impossible with `= 0`. This is a fix PR to make C++ compliant code by either adding missing virtual functions or sub-classing the abstract class.

### CODE REVIEW

**Comment by hcaseyal** [comment]
@veblush Could you rephrase the PR description? I'm not quite understanding it.

**Comment by veblush** [comment]
@hcaseyal I updated them to make it clearer :)

**Comment by veblush** [comment]
This should be enforced by a compiler and enabling a link to C++ standard library will solve this problem.

**Comment by hcaseyal** [comment]
@veblush Is there a plan to make this happen?

**Comment by veblush** [comment]
@hcaseyal #19918 is a on-going PR to enable C++ standard library and that is automatically enabled once that the PR is merged.
